In [1]:
import nearest_neighbors
import cross_validation
from sklearn.neighbors import NearestNeighbors
import sklearn.datasets
import matplotlib.pyplot as plt
import time
import re
import numpy as np
from sklearn.feature_extraction.text import TfidfTransformer
from scipy.sparse import csr_matrix
from sklearn.datasets import fetch_20newsgroups
from sklearn.metrics import confusion_matrix

In [8]:
#Task9,10,11
buf=fetch_20newsgroups(subset='train',remove=('headers', 'footers', 'quotes'))
ans=buf['target']
buf=buf['data']
#buf=[''.join((c if c.isalnum() else ' ') for c in s.lower()) for s in buf]
#buf2=[c.split() for c in buf]

In [3]:
#Task12
indptr = [0]
indices = []
data = []
vocabulary = {}
for d in buf2:
    for term in d:
        index = vocabulary.setdefault(term, len(vocabulary))
        indices.append(index)
        data.append(1)
    indptr.append(len(indices))
shape1=len(vocabulary)
shape0=len(buf2)
data1=csr_matrix((data, indices, indptr),shape=(shape0,shape1), dtype=float).toarray()

In [ ]:
#Task14
for k in ["euclidean","cosine"]:
    start=time.clock()
    a=cross_validation.knn_cross_val_score(data1,ans,[1,2,3,4,5,6,7,8,9,10],score="accuracy",cv=cross_validation.kfold(data1.shape[0],3),weights=False,metric=k,strategy="brute",test_block_size=20)
    print("time =",time.clock()-start)
    for i in a:
        mean=np.sum(a[i])/len(a[i])
        print("k =",i,k,"mean =",mean)

In [10]:
#Task13
transformer=TfidfTransformer()
tfidf = transformer.fit_transform(data1)

Мы убедились на предыдущей выборке, что точность на фолдах кросс-валидации алгоритмов "brute" и "my_own" совпадает, но имеет значительно разную скорость, поэтому на данной выборке лучше воспользоваться наиболее быстрым алгоритмом.

In [10]:
#Task14
for k in ["euclidean","cosine"]:
    start=time.clock()
    a=cross_validation.knn_cross_val_score(tfidf,ans,[1,2,3,4,5,6,7,8,9,10],score="accuracy",cv=cross_validation.kfold(tfidf.shape[0],3),weights=False,metric=k,strategy="brute",test_block_size=20)
    print("time =",time.clock()-start)
    for i in a:
        mean=np.sum(a[i])/len(a[i])
        print("k =",i,k,"mean =",mean)

time = 14.004056000000002
k = 1 euclidean mean = 0.153349162681
k = 2 euclidean mean = 0.1146362273
k = 3 euclidean mean = 0.107300191158
k = 4 euclidean mean = 0.103234306875
k = 5 euclidean mean = 0.098726546914
k = 6 euclidean mean = 0.0956328547409
k = 7 euclidean mean = 0.0937769815767
k = 8 euclidean mean = 0.0920094788613
k = 9 euclidean mean = 0.0890927384941
k = 10 euclidean mean = 0.085645962907
time = 11.759024000000004
k = 1 cosine mean = 0.582464158061
k = 2 cosine mean = 0.549230682632
k = 3 cosine mean = 0.552590236047
k = 4 cosine mean = 0.556832814359
k = 5 cosine mean = 0.559307022889
k = 6 cosine mean = 0.554622334557
k = 7 cosine mean = 0.551352159262
k = 8 cosine mean = 0.54993806804
k = 9 cosine mean = 0.547639709907
k = 10 cosine mean = 0.544723321053


In [10]:
#Task14
for k in ["euclidean","cosine"]:
    start=time.clock()
    a=cross_validation.knn_cross_val_score(tfidf,ans,[1,2,3,4,5,6,7,8,9,10],score="accuracy",cv=cross_validation.kfold(tfidf.shape[0],3),weights=True,metric=k,strategy="brute",test_block_size=20)
    print("time =",time.clock()-start)
    for i in a:
        mean=np.sum(a[i])/len(a[i])
        print("k =",i,k,"mean =",mean,"weights on")

time = 15.995240999999993
k = 1 euclidean mean = 0.157061986984 weights on
k = 2 euclidean mean = 0.156797367756 weights on
k = 3 euclidean mean = 0.151847661813 weights on
k = 4 euclidean mean = 0.14566079302 weights on
k = 5 euclidean mean = 0.141506444551 weights on
k = 6 euclidean mean = 0.134347243044 weights on
k = 7 euclidean mean = 0.126215708821 weights on
k = 8 euclidean mean = 0.117023377698 weights on
k = 9 euclidean mean = 0.113222628197 weights on
k = 10 euclidean mean = 0.110128959458 weights on
time = 14.093612000000007
k = 1 cosine mean = 0.589358271657 weights on
k = 2 cosine mean = 0.589269901208 weights on
k = 3 cosine mean = 0.589976630457 weights on
k = 4 cosine mean = 0.588032011896 weights on
k = 5 cosine mean = 0.583877804033 weights on
k = 6 cosine mean = 0.581403361161 weights on
k = 7 cosine mean = 0.575923338788 weights on
k = 8 cosine mean = 0.573006457815 weights on
k = 9 cosine mean = 0.570708591801 weights on
k = 10 cosine mean = 0.568587560422 weights 

Требовалось загрузить 20newsgroups, в которой документы, содержащие слова, преобразовать ее в разреженную матрицу и сравнить по кросс-валидации точность и скорость метода k ближайщих соседей при разных параметрах. Использовался метод "brute", нахваленный своей скоростью. Данные в таблице показывают точность. **Weights** означает взвешенный ли методы был, **Time** - время, за которое выполнил алгоритм, **Metric** - использовавшася метрика.

|Time| Weights | Metric | k=1 |k=2|k=3|k=4|k=5|k=6|k=7|k=8|k=9|k=10|
|:-----------: |:-----------:|:-----------:|:-----------:|:-----------:|:-----------:|:-----------:|:-----------:|:-----------:|:-----------:|:-----------:|:-----------:|:-----------:|
|14.00s| False       | Euclidean       | 0.1533       |0.1146|0.1073|0.1032|0.0987|0.0956|0.0937|0.0920|0.0890|0.0856|
|11.75s| False       | Cosine       | 0.5824       |0.5492|0.5525|0.5568|0.5593|0.5546|0.5513|0.5499|0.5476|0.5447|
|15.99s|   True     | Euclidean       | 0.1570       |0.1567|0.1518|0.1456|0.1415|0.1343|0.1262|0.1170|0.1132|0.1101|
|14.09s| True       | Cosine       | 0.5893     |0.5892|0.5899|0.5880|0.5838|0.5814|0.5759|0.5730|0.5707|0.5668|

Косинусная метрика показала контрастный результат в сравнении с евклидовой: выполнялась быстрее, точность заметно лучше. Поэтому на тестовой выборке 20newsgroups я использовал классификатор с параметрами: k=3, metric="cosine", weights=True.

In [4]:
#Task15
control2=fetch_20newsgroups(subset='test',remove=('headers', 'footers', 'quotes'))
control=control2['data']
ans2=control2['target']
control=[''.join((c if c.isalnum() else ' ') for c in s.lower()) for s in control]
test=[c.split() for c in control]

In [5]:
#Task15
indptr2 = [0]
indices2 = []
data2 = []
for d in test:
    for term in d:
        index2 = vocabulary.setdefault(term)
        if(index2!=None):
            indices2.append(index2)
            data2.append(1)
    indptr2.append(len(indices2))
shape2=len(test)
y_csr=csr_matrix((data2, indices2, indptr2), shape=(shape2,shape1),dtype=float).toarray()

In [11]:
#Task15
transformer2=TfidfTransformer()
test=transformer2.fit_transform(y_csr)

In [16]:
#Task15
classifier=nearest_neighbors.KNN_classifier(k=3,strategy='brute',metric='cosine',test_block_size=20,weights=True)
classifier.fit(tfidf,ans)
y_pred=classifier.predict(test)

In [17]:
#Task15
np.sum(y_pred==ans2)/len(ans2)

0.51035581518852891

Точность оказалась равной 0.5122, не очень близкой к кросс-валидационной.

In [18]:
#Task16
confusion_matrix(y_pred, ans2)

array([[156,   1,   5,   1,   1,   3,   3,   7,   7,   5,   3,   7,   9,
         23,  18,  63,  13,  39,  25,  43],
       [  0, 187,  17,  13,   8,  30,   3,   9,   3,   1,   0,   5,  12,
          8,   9,   0,   2,   1,   1,   0],
       [  3,  30, 190,  37,  16,  25,   9,   0,   7,   2,   3,   3,   9,
          1,   5,   0,   3,   0,   0,   1],
       [  2,  14,  32, 213,  45,   6,  42,   4,   6,   2,   0,   5,  18,
          4,   1,   1,   2,   3,   3,   2],
       [ 10,  15,  27,  38, 209,   8,  39,  26,  20,  16,  12,  20,  41,
         16,  25,  14,  11,   6,   9,   8],
       [  0,  36,  22,   7,   7, 263,   7,   1,   5,   2,   1,   2,  11,
          1,   3,   0,   2,   1,   1,   2],
       [  0,   6,  12,  14,   9,   5, 194,  12,   1,   4,   2,   4,   6,
          2,   1,   1,   3,   1,   3,   2],
       [  4,   5,   4,   7,   5,   1,  16, 206,  25,   5,   4,   2,  10,
          5,   8,   1,   6,   4,   4,   6],
       [  2,   7,   4,   1,   1,   2,   9,  19, 196,   5,   3,  

Значения на диагональных элемнтах довольно-таки невысокие, что говорит о не очень хорошем классификаторе. Много слов не было распознано, что показывают величины на внедиагональных элементах.

Вывел документы, на которых были допущены ошибки.

In [33]:
#Task16
print("Text 1:")
cur=np.where((ans2==1)&(ans2!=y_pred))[0]
np.random.shuffle(cur)
print(control2['data'][cur[1]])

print("\nText 2:")
cur=np.where((ans2==1)&(ans2!=y_pred))[0]
np.random.shuffle(cur)
print(control2['data'][cur[1]])

print("\nText 3:")
cur=np.where((ans2==1)&(ans2!=y_pred))[0]
np.random.shuffle(cur)
print(control2['data'][cur[1]])

Text 1:
Hello all,
   I need to make  some torso 3D scans and would like the phone numbers of
companies in the midwest that make scans, and the numbers of companies that
make the sanners (ie Cyberware). Does anyone have an idea of how much a
single scan costs and the best format to save it in? I am not sure on what
software platform I will be using it in, probably either Softimage or
Wavefront. So I think a spline based format would be best. Please forward the
numbers to me personally as I am having problems accessing USENET lately.
Thanks in advance!

Text 2:
I have a lots of .cgm files produced by NCAR Graph Utility V3.00.

They are all color graphs, and I want to print them out. 

The printers I own are Postscript, and HP 7475a. Anyone who have 

experiences in this please tell me, e-mail me will be very nice.

Or if someone knows how to convert those .cgm  files into .gif

pcx, .bmp .... , it will helps a lot. 


Text 3:
I'm told that VRrend386 is available on the internet. I wante

## Вывод

В обоих выборках косинусная метрика с взешенными весами показала лучший результат, нежели евклидовая. Это связно с тем, что она лучше работает в задачах распознавания рукописных цифр и схожести между текстами. К тому же в ней производится нормировка на длины векторов, благодаря чему она не зависит от размеров сравниваемых текстов. Поэтому при подобных задачах нужно пользоваться ей, а параметр k выбирать с помощью кросс-валидации.

## Список литературы

[Markdown Tutorial](http://www.markdowntutorial.com/)

[Найдется всё!](http://google.ru)